{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Association between covariates and PCs of gene expression data\n",
    "\n",
    "Same pipeline as `exp/PCA/ipynb/02_PCs_vs_covariates.ipynb` but using YARN-normalized GTEx RNA-Seq data.\n",
    "\n",
    "Most of the code blocks are identical to those in `exp/PCA/ipynb/02_PCs_vs_covariates.ipynb` except for \"Load expression data\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/json": {
       "Software versions": [
        {
         "module": "Python",
         "version": "2.7.12 64bit [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]"
        },
        {
         "module": "IPython",
         "version": "5.1.0"
        },
        {
         "module": "OS",
         "version": "Linux 2.6.32 431.3.1.el6.x86_64 x86_64 with centos 6.8 Final"
        },
        {
         "module": "numpy",
         "version": "1.11.2"
        },
        {
         "module": "pandas",
         "version": "0.19.1"
        },
        {
         "module": "sklearn",
         "version": "0.18.1"
        },
        {
         "module": "omics",
         "version": "0.1.4"
        }
       ]
      },
      "text/html": [
       "<table><tr><th>Software</th><th>Version</th></tr><tr><td>Python</td><td>2.7.12 64bit [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]</td></tr><tr><td>IPython</td><td>5.1.0</td></tr><tr><td>OS</td><td>Linux 2.6.32 431.3.1.el6.x86_64 x86_64 with centos 6.8 Final</td></tr><tr><td>numpy</td><td>1.11.2</td></tr><tr><td>pandas</td><td>0.19.1</td></tr><tr><td>sklearn</td><td>0.18.1</td></tr><tr><td>omics</td><td>0.1.4</td></tr><tr><td colspan='2'>Wed Jan 04 10:50:03 2017 EST</td></tr></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{|l|l|}\\hline\n",
       "{\\bf Software} & {\\bf Version} \\\\ \\hline\\hline\n",
       "Python & 2.7.12 64bit [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] \\\\ \\hline\n",
       "IPython & 5.1.0 \\\\ \\hline\n",
       "OS & Linux 2.6.32 431.3.1.el6.x86\\_64 x86\\_64 with centos 6.8 Final \\\\ \\hline\n",
       "numpy & 1.11.2 \\\\ \\hline\n",
       "pandas & 0.19.1 \\\\ \\hline\n",
       "sklearn & 0.18.1 \\\\ \\hline\n",
       "omics & 0.1.4 \\\\ \\hline\n",
       "\\hline \\multicolumn{2}{|l|}{Wed Jan 04 10:50:03 2017 EST} \\\\ \\hline\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "Software versions\n",
       "Python 2.7.12 64bit [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]\n",
       "IPython 5.1.0\n",
       "OS Linux 2.6.32 431.3.1.el6.x86_64 x86_64 with centos 6.8 Final\n",
       "numpy 1.11.2\n",
       "pandas 0.19.1\n",
       "sklearn 0.18.1\n",
       "omics 0.1.4\n",
       "Wed Jan 04 10:50:03 2017 EST"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from omics.stats.PCA import run_pca\n",
    "from omics.stats.MI import normalized_MI_matrix, MI\n",
    "\n",
    "%reload_ext version_information\n",
    "%version_information numpy, pandas, sklearn, omics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Input files from dbGaP\n",
    "EXPRESSION = \"/ifs/labs/cccb/projects/gtex/GTexV4/joey/src/hdf5/gtex_sub.hdf5\"\n",
    "\n",
    "# Data to use in this study\n",
    "PDATA       = \"../../PCA/data/pData.pickle\"\n",
    "GENE_LIST   = \"../../PCA/data/GTEx.genes_to_use.tsv\"\n",
    "SAMPLE_LIST = \"../../PCA/data/GTEx.samples_to_use.tsv\"\n",
    "\n",
    "# Ouptut files\n",
    "TISSUE_nMI = \"../reports/Tissue-specific_MI_between_PCs_and_covariates.xlsx\"  # based on YARN 30333 genes\n",
    "TISSUE_PCA = \"../reports/Tissue-specific_PCA_explained_variance_ratios.tsv\"   # based on YARN 30333 genes\n",
    "MAPPED_PCA = \"../data/samples_x_mapped_PCs.tsv\"  # based on YARN 30333 genes\n",
    "\n",
    "TISSUE_nMI2 = \"../reports/Tissue-specific_MI_between_PCs_and_covariates.n21088.xlsx\"  # YARN subset of 21088 genes in GENE_LIST\n",
    "TISSUE_PCA2 = \"../reports/Tissue-specific_PCA_explained_variance_ratios.n21088.tsv\"   # YARN subset of 21088 genes in GENE_LIST\n",
    "MAPPED_PCA2 = \"../data/samples_x_mapped_PCs.n21088.tsv\"  # YARN subset of 21088 genes in GENE_LIST\n",
    "\n",
    "# Others\n",
    "RND_SEED = 161126  # for calculating MI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load input data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8525\n"
     ]
    }
   ],
   "source": [
    "samples = pd.read_table(SAMPLE_LIST, index_col=0, squeeze=True)  # a series (SAMPID -> SMTS)\n",
    "print len(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30333, 8525)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hdf = pd.HDFStore(EXPRESSION)\n",
    "exprs = hdf['exprs'][samples.index]\n",
    "exprs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Test\n",
    "assert 'GTEX-1117F-0226-SM-5GZZ7' not in exprs  # not in data freeze due to Poor D-statistic\n",
    "assert 'GTEX-111CU-1826-SM-5GZYN' in exprs\n",
    "\n",
    "assert 'GTEX-YEC3-1426-101806-SM-5PNXX' not in exprs  # sample for DNA (SMNOTE), no sample-associated variables\n",
    "assert 'GTEX-YF7O-2326-101833-SM-5CVN9' not in exprs  # sample for DNA (SMNOTE), no sample-associated variables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load phenotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now we have 207 variables:\n",
      "category    151\n",
      "float64      56\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "pData = pd.read_pickle(PDATA)\n",
    "print 'Now we have', pData.shape[1], 'variables:'\n",
    "print pData.dtypes.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def impute_pData(pData):\n",
    "    return pData.fillna(pData.mean(numeric_only=True))\n",
    "\n",
    "def clean_pData(pData, cutoff=0.1):\n",
    "    n, k = pData.shape  # n samples x k variables\n",
    "    \n",
    "    is_enough = lambda x: len(x.dropna()) > n * (1-cutoff)  # %missing values < cutoff\n",
    "    is_stateful = lambda x: n > len(set(x.dropna())) > 1\n",
    "    is_informative = lambda x: MI(x, x) > cutoff  # information content (entropy) > cutoff\n",
    "    is_variant = lambda x: np.std(x) > 0\n",
    "    \n",
    "    pass1 = [i for i,x in pData.select_dtypes(['category']).iteritems() if is_enough(x) and is_stateful(x) and is_informative(x)]\n",
    "    pass2 = [i for i,x in pData.select_dtypes([np.number]).iteritems() if is_enough(x) and is_variant(x)]\n",
    "    print k, 'variables ->', len(pass1), 'categorical variables and', len(pass2), 'numeric variables.'\n",
    "        \n",
    "    return pData[sorted(pass1 + pass2)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tissue-specific covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def pc_vs_covariates_MI(exprs, pData, npc=10):\n",
    "    # Run PCA\n",
    "    pca, exprs_new = run_pca(exprs.T, pc=npc)\n",
    "    print \"Total explained variance of top %d PCs: %.2f\" % (npc, pca.explained_variance_ratio_.sum())\n",
    "    \n",
    "    # Joined feature dataframe (PCs + variables)\n",
    "    assert all(exprs_new.index == pData.index)  # check data integrity\n",
    "    df = pd.concat([exprs_new, pData], axis=1)\n",
    "    \n",
    "    # Compute normalized MI matrix\n",
    "    mat, nmat = normalized_MI_matrix(df, seed=RND_SEED, verbose=True)  # MI, nMI\n",
    "\n",
    "    return pca, exprs_new, mat, nmat\n",
    "\n",
    "\n",
    "def run_all_tissues(exprs, samples, pData, TISSUE_nMI, TISSUE_PCA, MAPPED_PCA):\n",
    "    xlsx = pd.ExcelWriter(TISSUE_nMI)  # output: tissue-specific MI matrix as a multi-sheet excel file\n",
    "    pca_var = dict()  # explained variance ratio of top n PCs in each tissue\n",
    "    pc_dfs = []  # mapped PC dataframes\n",
    "\n",
    "    for tissue in samples.astype('category').cat.categories:\n",
    "        sub_samples = samples[samples == tissue].index\n",
    "        sub_exprs = exprs[sub_samples]\n",
    "        sub_pData = pData.loc[sub_samples]\n",
    "        sub_pData2 = impute_pData(clean_pData(sub_pData))\n",
    "\n",
    "        print '================================'\n",
    "        print tissue\n",
    "        print '================================'\n",
    "        print 'Expression data:', sub_exprs.shape\n",
    "        print 'Phenotype data:', sub_pData.shape\n",
    "        print 'Imputed pData:', sub_pData2.shape\n",
    "\n",
    "        pca, exprs_new, mat, mat2 = pc_vs_covariates_MI(sub_exprs, sub_pData2)\n",
    "\n",
    "        pca_var[tissue] = pca.explained_variance_ratio_\n",
    "        pc_dfs.append(exprs_new)\n",
    "        mat2.to_excel(xlsx, tissue)\n",
    "        print\n",
    "\n",
    "    xlsx.save()\n",
    "    xlsx.close()\n",
    "    pd.DataFrame(pca_var).to_csv(TISSUE_PCA2, sep=\"\\t\")\n",
    "    pd.concat(pc_dfs).to_csv(MAPPED_PCA, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using YARN expression (n=30333)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "207 variables -> 98 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Adipose Tissue\n",
      "================================\n",
      "Expression data: (30333, 577)\n",
      "Phenotype data: (577, 207)\n",
      "Imputed pData: (577, 151)\n",
      "Data dimensions (samples-by-features): (577, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.12976429  0.08151015  0.03790651  0.03354869  0.02493951  0.02312883\n",
      "  0.01815523  0.01712829  0.01472719  0.0129264 ]\n",
      "Total explained variance of top 10 PCs: 0.39\n",
      "Input dataframe: 577 samples x 161 features\n",
      "98 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 88 categorical variables and 55 numeric variables.\n",
      "================================\n",
      "Adrenal Gland\n",
      "================================\n",
      "Expression data: (30333, 145)\n",
      "Phenotype data: (145, 207)\n",
      "Imputed pData: (145, 143)\n",
      "Data dimensions (samples-by-features): (145, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.13044141  0.07513339  0.04586832  0.02815608  0.02352348  0.02217572\n",
      "  0.01827786  0.01689078  0.01622014  0.01464933]\n",
      "Total explained variance of top 10 PCs: 0.39\n",
      "Input dataframe: 145 samples x 153 features\n",
      "88 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 143 categorical variables and 52 numeric variables.\n",
      "================================\n",
      "Blood\n",
      "================================\n",
      "Expression data: (30333, 511)\n",
      "Phenotype data: (511, 207)\n",
      "Imputed pData: (511, 195)\n",
      "Data dimensions (samples-by-features): (511, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.47275716  0.08794677  0.02221008  0.02011975  0.01292875  0.01191111\n",
      "  0.00756924  0.00683873  0.00654112  0.00638398]\n",
      "Total explained variance of top 10 PCs: 0.66\n",
      "Input dataframe: 511 samples x 205 features\n",
      "143 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 102 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Blood Vessel\n",
      "================================\n",
      "Expression data: (30333, 689)\n",
      "Phenotype data: (689, 207)\n",
      "Imputed pData: (689, 155)\n",
      "Data dimensions (samples-by-features): (689, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.14279993  0.07438337  0.04211626  0.02969735  0.02566218  0.02191855\n",
      "  0.0211101   0.01691022  0.01594396  0.01295109]\n",
      "Total explained variance of top 10 PCs: 0.40\n",
      "Input dataframe: 689 samples x 165 features\n",
      "102 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 79 categorical variables and 51 numeric variables.\n",
      "================================\n",
      "Brain\n",
      "================================\n",
      "Expression data: (30333, 1259)\n",
      "Phenotype data: (1259, 207)\n",
      "Imputed pData: (1259, 130)\n",
      "Data dimensions (samples-by-features): (1259, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.27782195  0.12308878  0.06183591  0.03254858  0.02805395  0.01944598\n",
      "  0.01714561  0.01426765  0.01194856  0.0107113 ]\n",
      "Total explained variance of top 10 PCs: 0.60\n",
      "Input dataframe: 1259 samples x 140 features\n",
      "79 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 88 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Breast\n",
      "================================\n",
      "Expression data: (30333, 214)\n",
      "Phenotype data: (214, 207)\n",
      "Imputed pData: (214, 141)\n",
      "Data dimensions (samples-by-features): (214, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.28421618  0.05470823  0.04073784  0.02645222  0.02183932  0.0202326\n",
      "  0.016551    0.01555841  0.01410418  0.01307894]\n",
      "Total explained variance of top 10 PCs: 0.51\n",
      "Input dataframe: 214 samples x 151 features\n",
      "88 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 91 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Colon\n",
      "================================\n",
      "Expression data: (30333, 345)\n",
      "Phenotype data: (345, 207)\n",
      "Imputed pData: (345, 144)\n",
      "Data dimensions (samples-by-features): (345, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.50281915  0.03637695  0.02794937  0.02299209  0.01446634  0.012439\n",
      "  0.01046633  0.00900057  0.00873264  0.00807507]\n",
      "Total explained variance of top 10 PCs: 0.65\n",
      "Input dataframe: 345 samples x 154 features\n",
      "91 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 87 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Esophagus\n",
      "================================\n",
      "Expression data: (30333, 686)\n",
      "Phenotype data: (686, 207)\n",
      "Imputed pData: (686, 140)\n",
      "Data dimensions (samples-by-features): (686, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.53952691  0.03830127  0.01768308  0.01260258  0.01156278  0.00981059\n",
      "  0.00928498  0.00855993  0.00776979  0.00743017]\n",
      "Total explained variance of top 10 PCs: 0.66\n",
      "Input dataframe: 686 samples x 150 features\n",
      "87 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 89 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Heart\n",
      "================================\n",
      "Expression data: (30333, 412)\n",
      "Phenotype data: (412, 207)\n",
      "Imputed pData: (412, 142)\n",
      "Data dimensions (samples-by-features): (412, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.18105547  0.07947732  0.05440741  0.02407892  0.01978918  0.01628698\n",
      "  0.01478303  0.01341563  0.0132433   0.01217448]\n",
      "Total explained variance of top 10 PCs: 0.43\n",
      "Input dataframe: 412 samples x 152 features\n",
      "89 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 83 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Kidney\n",
      "================================\n",
      "Expression data: (30333, 32)\n",
      "Phenotype data: (32, 207)\n",
      "Imputed pData: (32, 136)\n",
      "Data dimensions (samples-by-features): (32, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.17837373  0.14322206  0.09019419  0.06040133  0.04185165  0.03903352\n",
      "  0.03275194  0.03078135  0.02819894  0.02570685]\n",
      "Total explained variance of top 10 PCs: 0.67\n",
      "Input dataframe: 32 samples x 146 features\n",
      "83 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Liver\n",
      "================================\n",
      "Expression data: (30333, 119)\n",
      "Phenotype data: (119, 207)\n",
      "Imputed pData: (119, 126)\n",
      "Data dimensions (samples-by-features): (119, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.14483077  0.06228662  0.04454094  0.03144186  0.02659589  0.02319522\n",
      "  0.01989484  0.01758224  0.01646508  0.01577473]\n",
      "Total explained variance of top 10 PCs: 0.40\n",
      "Input dataframe: 119 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 86 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Lung\n",
      "================================\n",
      "Expression data: (30333, 320)\n",
      "Phenotype data: (320, 207)\n",
      "Imputed pData: (320, 139)\n",
      "Data dimensions (samples-by-features): (320, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.12043901  0.08208892  0.05782151  0.04148098  0.02806967  0.02235668\n",
      "  0.020172    0.01679089  0.0153254   0.01338282]\n",
      "Total explained variance of top 10 PCs: 0.42\n",
      "Input dataframe: 320 samples x 149 features\n",
      "86 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 100 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Muscle\n",
      "================================\n",
      "Expression data: (30333, 430)\n",
      "Phenotype data: (430, 207)\n",
      "Imputed pData: (430, 153)\n",
      "Data dimensions (samples-by-features): (430, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.09867697  0.04856283  0.03209326  0.02393986  0.02360534  0.02066829\n",
      "  0.01787902  0.01587173  0.01397359  0.01347407]\n",
      "Total explained variance of top 10 PCs: 0.31\n",
      "Input dataframe: 430 samples x 163 features\n",
      "100 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 92 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Nerve\n",
      "================================\n",
      "Expression data: (30333, 304)\n",
      "Phenotype data: (304, 207)\n",
      "Imputed pData: (304, 145)\n",
      "Data dimensions (samples-by-features): (304, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.08340636  0.05151257  0.03989653  0.03335104  0.02358212  0.02204792\n",
      "  0.02184864  0.02013284  0.01674028  0.01386439]\n",
      "Total explained variance of top 10 PCs: 0.33\n",
      "Input dataframe: 304 samples x 155 features\n",
      "92 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 79 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Ovary\n",
      "================================\n",
      "Expression data: (30333, 97)\n",
      "Phenotype data: (97, 207)\n",
      "Imputed pData: (97, 132)\n",
      "Data dimensions (samples-by-features): (97, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.10089669  0.064462    0.05345422  0.04173451  0.03396095  0.0275734\n",
      "  0.02576412  0.02160307  0.01975925  0.01869475]\n",
      "Total explained variance of top 10 PCs: 0.41\n",
      "Input dataframe: 97 samples x 142 features\n",
      "79 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 95 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Pancreas\n",
      "================================\n",
      "Expression data: (30333, 171)\n",
      "Phenotype data: (171, 207)\n",
      "Imputed pData: (171, 151)\n",
      "Data dimensions (samples-by-features): (171, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.06499524  0.04190651  0.03977966  0.02524768  0.02417955  0.02142725\n",
      "  0.02031087  0.01741004  0.01556873  0.01373485]\n",
      "Total explained variance of top 10 PCs: 0.28\n",
      "Input dataframe: 171 samples x 161 features\n",
      "95 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Pituitary\n",
      "================================\n",
      "Expression data: (30333, 103)\n",
      "Phenotype data: (103, 207)\n",
      "Imputed pData: (103, 126)\n",
      "Data dimensions (samples-by-features): (103, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.11060963  0.07252181  0.05419802  0.03433764  0.03297284  0.03050402\n",
      "  0.02476806  0.02099417  0.0192531   0.01769136]\n",
      "Total explained variance of top 10 PCs: 0.42\n",
      "Input dataframe: 103 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 66 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Prostate\n",
      "================================\n",
      "Expression data: (30333, 106)\n",
      "Phenotype data: (106, 207)\n",
      "Imputed pData: (106, 119)\n",
      "Data dimensions (samples-by-features): (106, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.12972723  0.07255552  0.06073552  0.04109725  0.03929176  0.02845823\n",
      "  0.02345971  0.02295641  0.02010616  0.01752327]\n",
      "Total explained variance of top 10 PCs: 0.46\n",
      "Input dataframe: 106 samples x 129 features\n",
      "66 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 66 categorical variables and 55 numeric variables.\n",
      "================================\n",
      "Salivary Gland\n",
      "================================\n",
      "Expression data: (30333, 57)\n",
      "Phenotype data: (57, 207)\n",
      "Imputed pData: (57, 121)\n",
      "Data dimensions (samples-by-features): (57, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.21824322  0.09942113  0.07102955  0.03958094  0.03275259  0.02964923\n",
      "  0.02606329  0.02348541  0.02093379  0.02047267]\n",
      "Total explained variance of top 10 PCs: 0.58\n",
      "Input dataframe: 57 samples x 131 features\n",
      "66 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 111 categorical variables and 52 numeric variables.\n",
      "================================\n",
      "Skin\n",
      "================================\n",
      "Expression data: (30333, 890)\n",
      "Phenotype data: (890, 207)\n",
      "Imputed pData: (890, 163)\n",
      "Data dimensions (samples-by-features): (890, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.57281886  0.02447506  0.02398543  0.0196233   0.01362326  0.01214093\n",
      "  0.00893864  0.00778285  0.0074654   0.00663935]\n",
      "Total explained variance of top 10 PCs: 0.70\n",
      "Input dataframe: 890 samples x 173 features\n",
      "111 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 90 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Small Intestine\n",
      "================================\n",
      "Expression data: (30333, 88)\n",
      "Phenotype data: (88, 207)\n",
      "Imputed pData: (88, 146)\n",
      "Data dimensions (samples-by-features): (88, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.28562427  0.15751657  0.0428043   0.03275966  0.02316997  0.02078682\n",
      "  0.01726611  0.01438442  0.01433304  0.01298929]\n",
      "Total explained variance of top 10 PCs: 0.62\n",
      "Input dataframe: 88 samples x 156 features\n",
      "90 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 108 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Spleen\n",
      "================================\n",
      "Expression data: (30333, 104)\n",
      "Phenotype data: (104, 207)\n",
      "Imputed pData: (104, 164)\n",
      "Data dimensions (samples-by-features): (104, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.07035035  0.0471344   0.0403496   0.03853076  0.03259713  0.03014302\n",
      "  0.02614519  0.02274685  0.02099126  0.01794502]\n",
      "Total explained variance of top 10 PCs: 0.35\n",
      "Input dataframe: 104 samples x 174 features\n",
      "108 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 84 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Stomach\n",
      "================================\n",
      "Expression data: (30333, 192)\n",
      "Phenotype data: (192, 207)\n",
      "Imputed pData: (192, 137)\n",
      "Data dimensions (samples-by-features): (192, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.41672219  0.04271744  0.03163056  0.02809844  0.01900367  0.01634852\n",
      "  0.01347301  0.01208299  0.01120138  0.00893488]\n",
      "Total explained variance of top 10 PCs: 0.60\n",
      "Input dataframe: 192 samples x 147 features\n",
      "84 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Testis\n",
      "================================\n",
      "Expression data: (30333, 172)\n",
      "Phenotype data: (172, 207)\n",
      "Imputed pData: (172, 126)\n",
      "Data dimensions (samples-by-features): (172, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.24260984  0.03429727  0.03334131  0.02742206  0.02280497  0.02088455\n",
      "  0.01691609  0.01500233  0.01271051  0.01223744]\n",
      "Total explained variance of top 10 PCs: 0.44\n",
      "Input dataframe: 172 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 83 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Thyroid\n",
      "================================\n",
      "Expression data: (30333, 323)\n",
      "Phenotype data: (323, 207)\n",
      "Imputed pData: (323, 136)\n",
      "Data dimensions (samples-by-features): (323, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.07289015  0.05336977  0.04533957  0.03560002  0.03012561  0.02474079\n",
      "  0.02138208  0.01720135  0.01552964  0.01523735]\n",
      "Total explained variance of top 10 PCs: 0.33\n",
      "Input dataframe: 323 samples x 146 features\n",
      "83 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 76 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Uterus\n",
      "================================\n",
      "Expression data: (30333, 83)\n",
      "Phenotype data: (83, 207)\n",
      "Imputed pData: (83, 129)\n",
      "Data dimensions (samples-by-features): (83, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.12114323  0.07686512  0.05184763  0.04239437  0.03046239  0.02900317\n",
      "  0.02568386  0.02292113  0.02036914  0.01762049]\n",
      "Total explained variance of top 10 PCs: 0.44\n",
      "Input dataframe: 83 samples x 139 features\n",
      "76 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 80 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Vagina\n",
      "================================\n",
      "Expression data: (30333, 96)\n",
      "Phenotype data: (96, 207)\n",
      "Imputed pData: (96, 133)\n",
      "Data dimensions (samples-by-features): (96, 30333)\n",
      "Variance explained by top 10 PCs: [ 0.3517355   0.05455892  0.04680414  0.02825131  0.02343482  0.01927755\n",
      "  0.01823693  0.0168061   0.01440172  0.01330632]\n",
      "Total explained variance of top 10 PCs: 0.59\n",
      "Input dataframe: 96 samples x 143 features\n",
      "80 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "run_all_tissues(exprs, samples, pData, TISSUE_nMI, TISSUE_PCA, MAPPED_PCA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intersection with GTEx-portal RPKM expression (n=21088)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21088, 8525)\n"
     ]
    }
   ],
   "source": [
    "genes = pd.read_table(GENE_LIST, index_col=0, squeeze=True)  # a series (ENSG ID -> Description)\n",
    "print len(genes), 'genes'  # 21146 genes\n",
    "exprs = exprs.loc[[i.split('.')[0] for i in genes.index]]  # remove version number in the Ensembl gene ID\n",
    "exprs = exprs.dropna()  # some samples are not in YARN-normalized expression profiles\n",
    "print exprs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "207 variables -> 98 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Adipose Tissue\n",
      "================================\n",
      "Expression data: (21088, 577)\n",
      "Phenotype data: (577, 207)\n",
      "Imputed pData: (577, 151)\n",
      "Data dimensions (samples-by-features): (577, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.15055937  0.09318292  0.04387981  0.0391703   0.02665359  0.02515286\n",
      "  0.02033328  0.01835177  0.01706872  0.0159765 ]\n",
      "Total explained variance of top 10 PCs: 0.45\n",
      "Input dataframe: 577 samples x 161 features\n",
      "98 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 88 categorical variables and 55 numeric variables.\n",
      "================================\n",
      "Adrenal Gland\n",
      "================================\n",
      "Expression data: (21088, 145)\n",
      "Phenotype data: (145, 207)\n",
      "Imputed pData: (145, 143)\n",
      "Data dimensions (samples-by-features): (145, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.15495104  0.08179164  0.05225373  0.03210874  0.02803493  0.02530738\n",
      "  0.02036338  0.01838142  0.01737991  0.01637349]\n",
      "Total explained variance of top 10 PCs: 0.45\n",
      "Input dataframe: 145 samples x 153 features\n",
      "88 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 143 categorical variables and 52 numeric variables.\n",
      "================================\n",
      "Blood\n",
      "================================\n",
      "Expression data: (21088, 511)\n",
      "Phenotype data: (511, 207)\n",
      "Imputed pData: (511, 195)\n",
      "Data dimensions (samples-by-features): (511, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.49686742  0.09027464  0.02354613  0.02184475  0.01525784  0.01255223\n",
      "  0.00874166  0.00801784  0.00737022  0.00702734]\n",
      "Total explained variance of top 10 PCs: 0.69\n",
      "Input dataframe: 511 samples x 205 features\n",
      "143 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 102 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Blood Vessel\n",
      "================================\n",
      "Expression data: (21088, 689)\n",
      "Phenotype data: (689, 207)\n",
      "Imputed pData: (689, 155)\n",
      "Data dimensions (samples-by-features): (689, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.17246065  0.08235337  0.04514199  0.03322226  0.02756082  0.02443281\n",
      "  0.02249283  0.0180762   0.01659454  0.01443723]\n",
      "Total explained variance of top 10 PCs: 0.46\n",
      "Input dataframe: 689 samples x 165 features\n",
      "102 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 79 categorical variables and 51 numeric variables.\n",
      "================================\n",
      "Brain\n",
      "================================\n",
      "Expression data: (21088, 1259)\n",
      "Phenotype data: (1259, 207)\n",
      "Imputed pData: (1259, 130)\n",
      "Data dimensions (samples-by-features): (1259, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.30398803  0.13772171  0.06425461  0.03581296  0.02990555  0.02239135\n",
      "  0.01886526  0.01566987  0.01290451  0.01174694]\n",
      "Total explained variance of top 10 PCs: 0.65\n",
      "Input dataframe: 1259 samples x 140 features\n",
      "79 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 88 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Breast\n",
      "================================\n",
      "Expression data: (21088, 214)\n",
      "Phenotype data: (214, 207)\n",
      "Imputed pData: (214, 141)\n",
      "Data dimensions (samples-by-features): (214, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.32832199  0.06158984  0.04326066  0.02794946  0.02294713  0.021172\n",
      "  0.01802646  0.01576706  0.01498771  0.01343786]\n",
      "Total explained variance of top 10 PCs: 0.57\n",
      "Input dataframe: 214 samples x 151 features\n",
      "88 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 91 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Colon\n",
      "================================\n",
      "Expression data: (21088, 345)\n",
      "Phenotype data: (345, 207)\n",
      "Imputed pData: (345, 144)\n",
      "Data dimensions (samples-by-features): (345, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.56900073  0.03607897  0.03004765  0.02557219  0.01499312  0.01225323\n",
      "  0.01065438  0.00916091  0.00872766  0.0075875 ]\n",
      "Total explained variance of top 10 PCs: 0.72\n",
      "Input dataframe: 345 samples x 154 features\n",
      "91 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 87 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Esophagus\n",
      "================================\n",
      "Expression data: (21088, 686)\n",
      "Phenotype data: (686, 207)\n",
      "Imputed pData: (686, 140)\n",
      "Data dimensions (samples-by-features): (686, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.60763497  0.03688745  0.01899833  0.01303781  0.01134597  0.00971391\n",
      "  0.00917173  0.00826819  0.00756437  0.00714501]\n",
      "Total explained variance of top 10 PCs: 0.73\n",
      "Input dataframe: 686 samples x 150 features\n",
      "87 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 89 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Heart\n",
      "================================\n",
      "Expression data: (21088, 412)\n",
      "Phenotype data: (412, 207)\n",
      "Imputed pData: (412, 142)\n",
      "Data dimensions (samples-by-features): (412, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.21176939  0.08821902  0.06215487  0.02618387  0.02255562  0.01835198\n",
      "  0.01659212  0.01561438  0.01481238  0.01362656]\n",
      "Total explained variance of top 10 PCs: 0.49\n",
      "Input dataframe: 412 samples x 152 features\n",
      "89 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 83 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Kidney\n",
      "================================\n",
      "Expression data: (21088, 32)\n",
      "Phenotype data: (32, 207)\n",
      "Imputed pData: (32, 136)\n",
      "Data dimensions (samples-by-features): (32, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.20185458  0.15755492  0.0942063   0.06474398  0.04111115  0.03841083\n",
      "  0.03292039  0.03007504  0.02658557  0.02405147]\n",
      "Total explained variance of top 10 PCs: 0.71\n",
      "Input dataframe: 32 samples x 146 features\n",
      "83 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Liver\n",
      "================================\n",
      "Expression data: (21088, 119)\n",
      "Phenotype data: (119, 207)\n",
      "Imputed pData: (119, 126)\n",
      "Data dimensions (samples-by-features): (119, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.16342646  0.07028496  0.05282485  0.03238403  0.02888413  0.02542503\n",
      "  0.0222738   0.0201063   0.0179396   0.01693957]\n",
      "Total explained variance of top 10 PCs: 0.45\n",
      "Input dataframe: 119 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 86 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Lung\n",
      "================================\n",
      "Expression data: (21088, 320)\n",
      "Phenotype data: (320, 207)\n",
      "Imputed pData: (320, 139)\n",
      "Data dimensions (samples-by-features): (320, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.13543269  0.08821546  0.06817221  0.04494161  0.03432465  0.02616808\n",
      "  0.02323434  0.01969724  0.01825349  0.01590034]\n",
      "Total explained variance of top 10 PCs: 0.47\n",
      "Input dataframe: 320 samples x 149 features\n",
      "86 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 100 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Muscle\n",
      "================================\n",
      "Expression data: (21088, 430)\n",
      "Phenotype data: (430, 207)\n",
      "Imputed pData: (430, 153)\n",
      "Data dimensions (samples-by-features): (430, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.11496776  0.05690174  0.03551064  0.02793047  0.02612511  0.02515521\n",
      "  0.01974144  0.01809532  0.01675941  0.01529925]\n",
      "Total explained variance of top 10 PCs: 0.36\n",
      "Input dataframe: 430 samples x 163 features\n",
      "100 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 92 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Nerve\n",
      "================================\n",
      "Expression data: (21088, 304)\n",
      "Phenotype data: (304, 207)\n",
      "Imputed pData: (304, 145)\n",
      "Data dimensions (samples-by-features): (304, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.09991669  0.06204209  0.04096763  0.03863013  0.02828066  0.02710193\n",
      "  0.02621084  0.02484216  0.01992677  0.01525044]\n",
      "Total explained variance of top 10 PCs: 0.38\n",
      "Input dataframe: 304 samples x 155 features\n",
      "92 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 79 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Ovary\n",
      "================================\n",
      "Expression data: (21088, 97)\n",
      "Phenotype data: (97, 207)\n",
      "Imputed pData: (97, 132)\n",
      "Data dimensions (samples-by-features): (97, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.11878004  0.073214    0.05822212  0.04616939  0.03653185  0.02912506\n",
      "  0.02867573  0.02361327  0.02114018  0.02051349]\n",
      "Total explained variance of top 10 PCs: 0.46\n",
      "Input dataframe: 97 samples x 142 features\n",
      "79 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 95 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Pancreas\n",
      "================================\n",
      "Expression data: (21088, 171)\n",
      "Phenotype data: (171, 207)\n",
      "Imputed pData: (171, 151)\n",
      "Data dimensions (samples-by-features): (171, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.08115289  0.05047853  0.04698076  0.02964235  0.02856413  0.02588888\n",
      "  0.02324731  0.02142118  0.0185523   0.01577114]\n",
      "Total explained variance of top 10 PCs: 0.34\n",
      "Input dataframe: 171 samples x 161 features\n",
      "95 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Pituitary\n",
      "================================\n",
      "Expression data: (21088, 103)\n",
      "Phenotype data: (103, 207)\n",
      "Imputed pData: (103, 126)\n",
      "Data dimensions (samples-by-features): (103, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.13416596  0.08746089  0.05903084  0.03861447  0.03547753  0.03275282\n",
      "  0.02649629  0.02317668  0.0204918   0.01928885]\n",
      "Total explained variance of top 10 PCs: 0.48\n",
      "Input dataframe: 103 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 66 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Prostate\n",
      "================================\n",
      "Expression data: (21088, 106)\n",
      "Phenotype data: (106, 207)\n",
      "Imputed pData: (106, 119)\n",
      "Data dimensions (samples-by-features): (106, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.14502086  0.08072514  0.06880299  0.04716833  0.04430418  0.03221543\n",
      "  0.02579884  0.02506148  0.02143014  0.01877856]\n",
      "Total explained variance of top 10 PCs: 0.51\n",
      "Input dataframe: 106 samples x 129 features\n",
      "66 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 66 categorical variables and 55 numeric variables.\n",
      "================================\n",
      "Salivary Gland\n",
      "================================\n",
      "Expression data: (21088, 57)\n",
      "Phenotype data: (57, 207)\n",
      "Imputed pData: (57, 121)\n",
      "Data dimensions (samples-by-features): (57, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.25377616  0.11923689  0.07795891  0.04196067  0.03583188  0.03028558\n",
      "  0.02537583  0.02362233  0.02094318  0.01917849]\n",
      "Total explained variance of top 10 PCs: 0.65\n",
      "Input dataframe: 57 samples x 131 features\n",
      "66 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 111 categorical variables and 52 numeric variables.\n",
      "================================\n",
      "Skin\n",
      "================================\n",
      "Expression data: (21088, 890)\n",
      "Phenotype data: (890, 207)\n",
      "Imputed pData: (890, 163)\n",
      "Data dimensions (samples-by-features): (890, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.63261519  0.02590343  0.02411726  0.02101722  0.0134176   0.01251811\n",
      "  0.00860442  0.00793496  0.00733424  0.00631731]\n",
      "Total explained variance of top 10 PCs: 0.76\n",
      "Input dataframe: 890 samples x 173 features\n",
      "111 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 90 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Small Intestine\n",
      "================================\n",
      "Expression data: (21088, 88)\n",
      "Phenotype data: (88, 207)\n",
      "Imputed pData: (88, 146)\n",
      "Data dimensions (samples-by-features): (88, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.33346066  0.17006484  0.04750551  0.03279067  0.0249886   0.02154438\n",
      "  0.01743793  0.01530508  0.01400296  0.01296687]\n",
      "Total explained variance of top 10 PCs: 0.69\n",
      "Input dataframe: 88 samples x 156 features\n",
      "90 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 108 categorical variables and 56 numeric variables.\n",
      "================================\n",
      "Spleen\n",
      "================================\n",
      "Expression data: (21088, 104)\n",
      "Phenotype data: (104, 207)\n",
      "Imputed pData: (104, 164)\n",
      "Data dimensions (samples-by-features): (104, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.08220557  0.05251893  0.04821946  0.04327011  0.03896394  0.03418244\n",
      "  0.02900471  0.02623704  0.02448343  0.0202222 ]\n",
      "Total explained variance of top 10 PCs: 0.40\n",
      "Input dataframe: 104 samples x 174 features\n",
      "108 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 84 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Stomach\n",
      "================================\n",
      "Expression data: (21088, 192)\n",
      "Phenotype data: (192, 207)\n",
      "Imputed pData: (192, 137)\n",
      "Data dimensions (samples-by-features): (192, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.47982408  0.04736414  0.03231816  0.02972678  0.02012762  0.0175831\n",
      "  0.01434315  0.01269033  0.01163899  0.00897012]\n",
      "Total explained variance of top 10 PCs: 0.67\n",
      "Input dataframe: 192 samples x 147 features\n",
      "84 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 73 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Testis\n",
      "================================\n",
      "Expression data: (21088, 172)\n",
      "Phenotype data: (172, 207)\n",
      "Imputed pData: (172, 126)\n",
      "Data dimensions (samples-by-features): (172, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.28515517  0.04254505  0.03964074  0.03285805  0.02688594  0.02220483\n",
      "  0.01875568  0.01599366  0.01354279  0.01276255]\n",
      "Total explained variance of top 10 PCs: 0.51\n",
      "Input dataframe: 172 samples x 136 features\n",
      "73 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 83 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Thyroid\n",
      "================================\n",
      "Expression data: (21088, 323)\n",
      "Phenotype data: (323, 207)\n",
      "Imputed pData: (323, 136)\n",
      "Data dimensions (samples-by-features): (323, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.08389207  0.06986296  0.05273564  0.0362203   0.03376109  0.03096837\n",
      "  0.02604316  0.01986041  0.01821492  0.01692901]\n",
      "Total explained variance of top 10 PCs: 0.39\n",
      "Input dataframe: 323 samples x 146 features\n",
      "83 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 76 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Uterus\n",
      "================================\n",
      "Expression data: (21088, 83)\n",
      "Phenotype data: (83, 207)\n",
      "Imputed pData: (83, 129)\n",
      "Data dimensions (samples-by-features): (83, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.1330035   0.08840319  0.05623609  0.04753176  0.03409946  0.0324564\n",
      "  0.02705949  0.02483221  0.02106045  0.01931951]\n",
      "Total explained variance of top 10 PCs: 0.48\n",
      "Input dataframe: 83 samples x 139 features\n",
      "76 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n",
      "207 variables -> 80 categorical variables and 53 numeric variables.\n",
      "================================\n",
      "Vagina\n",
      "================================\n",
      "Expression data: (21088, 96)\n",
      "Phenotype data: (96, 207)\n",
      "Imputed pData: (96, 133)\n",
      "Data dimensions (samples-by-features): (96, 21088)\n",
      "Variance explained by top 10 PCs: [ 0.40819409  0.05438998  0.05285488  0.03043271  0.02350292  0.02038629\n",
      "  0.01887001  0.0168647   0.01464313  0.01287421]\n",
      "Total explained variance of top 10 PCs: 0.65\n",
      "Input dataframe: 96 samples x 143 features\n",
      "80 features are categorical\n",
      "Computing pair-wise MI ...\n",
      "Normalize MI ...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "run_all_tissues(exprs, samples, pData, TISSUE_nMI2, TISSUE_PCA2, MAPPED_PCA2)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "toc": {
   "nav_menu": {
    "height": "12px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
